### PR TITLE
Data storage objects

### LINKED ISSUE #90: Consistency in Rows/Columns/1d-arrays
It seems like in general when we have tasks requiring linear algebra operations on arrays, it is good practice to ensure that we have a defined dimensionality of the arrays. **I would recommend we always use 2D arrays** (columns or rows) over 1D vector-type arrays if we are going to use them for linear algebra. Typically these arising from a 'vanishing dimension' of a normally clear 2D array.

E.g when dealing with ensemble based methods, we perhaps wish for arrays to be `[ensemble size x parameter dimension]` but if we have parameter dimension of 1 then often one defaults to a 1D-array, from which the ambiguity of "which dimension vanished?" arises. This is cured by specifying 2D inputs of  `[ensemble size x 1]` or a `[1 x parameter dimension]`.

The overhead then to pick the dimensions consistently throughout CES to avoid multiple transpositions as we pass through different objects or functions. 

*note to self, Julia encourages the use of `permutedims(A,(2,1))` over `A'` for applying a transposition otherwise one changes the type into an Adjoint LinearAlgebra object.*

### PR DESCRIPTION
**Pull request by odunbar**
## Purpose
Fixes #90
This PR adds `DataStorage.jl` containing objects to store singleton, or paired, data samples. The idea is to help ensure dimensionally consistent data (as these **always store a 2D array where the samples are columns**, so if in doubt - the data are columns), which has lead to numerous issues in the past. They also ensure that your input-output pairs for the emulator are safeguarded (e.g it checks consistent dimensions of samples). 

Users will need to specify orientation when they provide arrays (as to whether the data stored as a list of rows, or a list of columns), then internally they are arranged always as columns. Effectively therefore we are really just adding more rigid constructors for data arrays.

## Contains:
 - The `DataContainer` object stores data as columns
 - The `PairedDataContainer` object stores 2 `DataContainer` objects, one for input data, one for output data. The number of samples is checked to be the same length in inputs and outputs, while the data dimensions can differ.
 - Unit tests for the object and related functions, also updated unit tests for the other objects to use these
 - Small changes to examples (Lorenz 96/Cloudy/LossMinimization/GaussianProcessEmulator) to work with new data storage 
 - To obtain training points from the EKP: in `src/Utilities` there is a `get_training_points` function which returns a `PairedDataContainer`
 - The GaussianProcessEmulator only takes in `PairedDataContainer` now
### Also...
 - I required some larger changes to the Cloudy example run (even without the new data storage))
 - I removed some confusing `vcat([X.first, X.last]...)` type lines in GP setup and replaced with interpretable loop
 - getters and setters for EKP, I really push for us to use getters/setters as an interface
 - MCMC stores posterior array with data-as-columns, and utilities work with this too.

### DataContainer
To create a Data container simply create an array, and specify whether you consider data within as a list of columns, or not. e.g for 2D data and 10 samples
```julia
x = rand(2,10)
column_store = DataContainer(x,data_are_columns=true)
```
or for 10D data and 2 samples
```julia
column_store2 = DataContainer(x,data_are_columns=false)
```
Remember, the data is **always** stored as columns so
```julia
@assert(get_data(column_store) == get_data(column_store2))
```
returns `true`
### PairedDataContainer
Can be constructed from 2 `DataContainer`s of the same sample size.
```julia
input_output_pairs = PairedDataContainer(column_store,column_store)
```
Or from arrays themselves, (so long as they have the same "samples" dimension)
```julia
input_output_pairs2 = PairedDataContainer(x,x,data_are_columns=true) 
```
and, 
```julia
PairedDataContainer(column_store,column_store2) 
```
throws a dimension mismatch error.

### CODE REVIEW

File: src/EnsembleKalmanProcesses.jl [hunk 1]
```diff
@@ -46,89 +48,125 @@ Structure that is used in Ensemble Kalman processes
...
 end
 
 
+
+"""
+    get_X(ekp::EnsembleKalmanProcess, iteration::IT; return_array=true) where {IT <: Integer}
+
+Get X=u or X=g for the EKI iteration. Returns a DataContainer object unless array is specified.
+"""
+function get_u(ekp::EnsembleKalmanProcess, iteration::IT; return_array=true) where {IT <: Integer}
```
Comment (ilopezgp) [inline src/EnsembleKalmanProcesses.jl:98]: We might want to have a getter for the whole sequence. E.g., `get_u(ekp)` that returns an array or list of Array{FT,2}.

File: src/DataStorage.jl [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+#other constructors
+
+
+## functions
+function size(dc::DataContainer)
+    return size(dc.stored_data)
+end
+
+function size(dc::DataContainer,idx::IT) where {IT <: Integer}
+    return size(dc.stored_data,idx)
```
[the remark this answers is not in the document]
Comment (odunbar) [reply src/DataStorage.jl:71]: Would you prefer just `idx::Int64` ? I am still a bit unsure about the Julia type preferences - I've seen quite a lot of the following style:
`x::X where {X <: AbstractType}`
which feels less likely to cause errors, as it allows any "form" of integer. i'm happy to switch to an alternative

File: src/EnsembleKalmanProcesses.jl [hunk 3]
```diff
@@ -46,89 +48,125 @@ Structure that is used in Ensemble Kalman processes
...
-     process::P
+    "Array of stores for parameters (u), each of size [N_data x N_ensemble]"
+    u::Array{DataContainer{FT}}
+    "vector of the observed vector size [N_data]"
+    obs_mean::Vector{FT}
+    "covariance of the observational noise, which is assumed to be normally distributed"
+    obs_noise_cov::Array{FT, 2}
+    "ensemble size"
+    N_ens::IT
+    "Array of stores for forward model outputs, each of size  [N_data x N_ensembles]; mean of all observation samples"
```
[the remark this answers is not in the document]
Comment (odunbar) [reply src/EnsembleKalmanProcesses.jl:59]: This is a typo and has been removed. I'm also going to change `N_data` to `data_dim` as it is ambiguous.

File: src/DataStorage.jl [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+## Objects
+"""
+    struct DataContainer
+
+struct to store data samples as columns in an array
+"""
+struct DataContainer{FT <: Real}
+    #stored data, each piece of data is a column [number of samples x data size]
+    stored_data::Array{FT,2}    
+    DataContainer(stored_data::Array{FT,2}; data_are_columns=true) where {FT <: Real} = data_are_columns ? new{FT}(stored_data) : new{FT}(permutedims(stored_data,(2,1)))
```
[the remark this answers is not in the document]
Comment (odunbar) [reply src/DataStorage.jl:20]: Done

File: src/EnsembleKalmanProcesses.jl (hunk 1 above)
Comment (odunbar) [reply src/EnsembleKalmanProcesses.jl:99]: Done. I have implemented this to returns a list of DataContainers or a list of arrays. Along with some runtests.

File: src/DataStorage.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (ilopezgp) [reply src/DataStorage.jl:83]: Let's leave it then

File: src/DataStorage.jl [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+export get_data, get_inputs, get_outputs
+
+## Objects
+"""
+    struct DataContainer
+
+struct to store data samples as columns in an array
+"""
+struct DataContainer{FT <: Real}
+    #stored data, each piece of data is a column [number of samples x data size]
```
Comment (ilopezgp) [inline src/DataStorage.jl:18]: Wouldn't `stored_data` be an array `[number of samples x data size]`, where number of samples is the size of the ensemble and data size is the number of parameters?

File: src/DataStorage.jl [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    end
+end
+
+"""
+    PairedDataContainer
+
+stores input - output pairs as data containers
+"""
+struct PairedDataContainer{FT <: Real}
+    #container for inputs and ouputs [sample_size x data/parameter size]
```
Comment (ilopezgp) [inline src/DataStorage.jl:39]: Same here

File: src/EnsembleKalmanProcesses.jl [hunk 7]
```diff
@@ -46,89 +48,137 @@ Structure that is used in Ensemble Kalman processes
...
     Random.seed!(rng_seed)
-    params = permutedims(sample_distribution(prior, N_ens), (2,1)) #this transpose is [N_ens x dim(param space)]
-    return params
+    parameters = sample_distribution(prior, N_ens) #of size [dim(param space) N_ens]
+    return parameters
 end
 
 function compute_error!(ekp::EnsembleKalmanProcess)
-    mean_g = dropdims(mean(ekp.g[end], dims=1), dims=1)
+    mean_g = dropdims(mean(get_data(ekp.g[end]), dims=2), dims=2)
```
Comment (ilopezgp) [inline src/EnsembleKalmanProcesses.jl:153]: Looks like a good place to use `get_g`

File: src/EnsembleKalmanProcesses.jl [hunk 8]
```diff
@@ -208,26 +277,34 @@ function update_ensemble!(ekp::EnsembleKalmanProcess{FT, IT, Inversion}, g; cov_
...
+    # g: N_ens × data_dim
+    u_old = get_u_final(ekp)
+    u_old = permutedims(u_old,(2,1))
+    u = u_old
+    g = permutedims(g_in, (2,1))
+   
+    # u_mean: parameter_dim × 1
     u_mean = mean(u', dims=2)
-    # g_mean: N_params x 1
+    # g_mean: parameter_dim × 1
```
Comment (ilopezgp) [inline src/EnsembleKalmanProcesses.jl:296]: `data_dim x 1`

File: src/EnsembleKalmanProcesses.jl [hunk 9]
```diff
@@ -208,26 +277,34 @@ function update_ensemble!(ekp::EnsembleKalmanProcess{FT, IT, Inversion}, g; cov_
...
     g_mean = mean(g', dims=2)
-    # g_cov: N_params x N_params
+    # g_cov: parameter_dim × parameter_dim
```
Comment (ilopezgp) [inline src/EnsembleKalmanProcesses.jl:298]: `data_dim x data_dim`

File: src/GaussianProcessEmulator.jl [hunk 10]
```diff
@@ -56,36 +57,33 @@ $(DocStringExtensions.FIELDS)
...
-    inputs::Array{FT}
-    "training data/targets; N_samples x output_dim"
-    data::Array{FT}
+    "training inputs and outputs, data stored in columns"
+    input_output_pairs::PairedDataContainer{FT}
     "mean of input; 1 x input_dim"
     input_mean::Array{FT}
     "the Gaussian Process (GP) Regression model(s) that are fitted to the given input-data pairs"
+    sqrt_inv_input_cov::Union{Nothing, Array{FT, 2}}
+    "prediction type (`y` to predict the data, `f` to predict the latent function)"
```
Comment (ilopezgp) [inline src/GaussianProcessEmulator.jl:66]: I think the descriptions here do not match (prediction type and sqrt_inv_input_cov are switched)

File: src/GaussianProcessEmulator.jl [hunk 11]
```diff
@@ -56,36 +57,33 @@ $(DocStringExtensions.FIELDS)
...
     "square root of the inverse of the input covariance matrix; input_dim x input_dim"
-    sqrt_inv_input_cov::Union{Nothing, Array{FT, 2}}
-    "prediction type (`y` to predict the data, `f` to predict the latent function)"
     prediction_type::Union{Nothing, PredictionType}
 end
 
 
 """
-GaussianProcess(inputs::Array{FT, 2}, data::Array{FT, 2}, package::GPJL; GPkernel::Union{K, KPy, Nothing}=nothing, obs_noise_cov::Union{Array{FT, 2}, Nothing}=nothing, normalized::Bool=true, noise_learn::Bool=true, prediction_type::PredictionType=YType()) where {K<:Kernel, KPy<:PyObject}
+    GaussianProcess(inputs::Array{FT, 2}, data::Array{FT, 2}, package::GPJL; GPkernel::Union{K, KPy, Nothing}=nothing, obs_noise_cov::Union{Array{FT, 2}, Nothing}=nothing, normalized::Bool=true, noise_learn::Bool=true, prediction_type::PredictionType=YType()) where {K<:Kernel, KPy<:PyObject}
```
Comment (ilopezgp) [inline src/GaussianProcessEmulator.jl:78]: Outdated description, does not contain `input_output_pairs::PairedDataContainer{FT},`

File: src/GaussianProcessEmulator.jl [hunk 12]
```diff
@@ -192,29 +181,26 @@ function GaussianProcess(
...
+                                               prediction_type)
 end
 
 
 """
-GaussianProcess(inputs::Array{FT, 2}, data::Array{FT, 2}, package::SKLJL; GPkernel::Union{K, KPy, Nothing}=nothing, obs_noise_cov::Union{Array{FT, 2}, Nothing}=nothing, normalized::Bool=true, noise_learn::Bool=true, prediction_type::PredictionType=YType()) where {K<:Kernel, KPy<:PyObject}
-
-
-Inputs and data of size N_samples x input_dim
+    GaussianProcess(inputs::Array{FT, 2}, data::Array{FT, 2}, package::SKLJL; GPkernel::Union{K, KPy, Nothing}=nothing, obs_noise_cov::Union{Array{FT, 2}, Nothing}=nothing, normalized::Bool=true, noise_learn::Bool=true, prediction_type::PredictionType=YType()) where {K<:Kernel, KPy<:PyObject}
```
Comment (ilopezgp) [inline src/GaussianProcessEmulator.jl:196]: Outdated description, include `input_output_pairs::PairedDataContainer{FT},`

File: src/GaussianProcessEmulator.jl [hunk 13]
```diff
@@ -340,45 +320,38 @@ function predict(gp::GaussianProcess{FT, GPJL}, new_inputs::Array{FT, 2}, transf
...
     # dimension. 
-    # new_inputs should be of size N_new_inputs x input_dim, so input_dim
-    # has to equal the input dimension of the GP model(s). 
-    input_dim = size(gp.inputs, 2)
-    output_dim = size(gp.data, 2)
-    N_new_inputs = size(new_inputs, 1)
-    err = "GP object and input observations do not have consistent dimensions"
-    size(new_inputs, 2) == input_dim || throw(ArgumentError(err))
+    input_dim, output_dim = size(gp.input_output_pairs, 1)
+    N_new_inputs = size(new_inputs, 2)
```
Comment (ilopezgp) [inline src/GaussianProcessEmulator.jl:324]: I think now `N_new_inputs` should be renamed to `N_samples`

File: test/DataStorage/runtests.jl [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+using Distributions
+using Random
+
+using CalibrateEmulateSample.DataStorage
+
+@testset "DataStorage" begin
+    seed=2021
+    Random.seed!(seed)
+        
+    parameter_samples = rand(MvNormal(2,0.1),10) #10 samples of 4D params
```
Comment (ilopezgp) [inline test/DataStorage/runtests.jl:11]: 2D samples

File: test/DataStorage/runtests.jl [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    data_samples = rand(MvNormal(12,2),10) #10 samples of 12D data
+    data_samples_short = data_samples[:,1:end-1]
+
+    new_parameter_samples = rand(MvNormal(2,0.1),10) #10 samples of 4D params
```
Comment (ilopezgp) [inline test/DataStorage/runtests.jl:15]: 2D

File: test/MarkovChainMonteCarlo/runtests.jl [hunk 16]
```diff
@@ -17,9 +18,11 @@ using CalibrateEmulateSample.GaussianProcessEmulator
     # We need a GaussianProcess to run MarkovChainMonteCarlo, so let's reconstruct the one that's tested
     # in test/GaussianProcesses/runtests.jl
     n = 20                                       # number of training points
-    x = reshape(2.0 * π * rand(n), n, 1)         # predictors/features: n x 1
-    σ2_y = reshape([0.05], 1, 1)
-    y = reshape(sin.(x) + rand(Normal(0, σ2_y[1]), n), n, 1) # predictands/targets: n x 1
+    x = reshape(2.0 * π * rand(n), 1, n)         # predictors/features: n x 1
```
Comment (ilopezgp) [inline test/MarkovChainMonteCarlo/runtests.jl:21]: `# predictors/features: 1 x n`

File: src/DataStorage.jl (hunk 5 above)
Comment (odunbar) [reply src/DataStorage.jl:18]: should be [data_dim x n_samples]. Changed!

File: src/DataStorage.jl (hunk 6 above)
Comment (odunbar) [reply src/DataStorage.jl:39]: Done

File: src/EnsembleKalmanProcesses.jl (hunk 7 above)
Comment (odunbar) [reply src/EnsembleKalmanProcesses.jl:153]: Done

File: src/EnsembleKalmanProcesses.jl (hunk 8 above)
Comment (odunbar) [reply src/EnsembleKalmanProcesses.jl:296]: Done

File: src/EnsembleKalmanProcesses.jl (hunk 9 above)
Comment (odunbar) [reply src/EnsembleKalmanProcesses.jl:298]: Done

File: test/DataStorage/runtests.jl (hunk 14 above)
Comment (odunbar) [reply test/DataStorage/runtests.jl:11]: Done

File: test/MarkovChainMonteCarlo/runtests.jl (hunk 16 above)
Comment (odunbar) [reply test/MarkovChainMonteCarlo/runtests.jl:21]: Done

File: src/GaussianProcessEmulator.jl (hunk 10 above)
Comment (odunbar) [reply src/GaussianProcessEmulator.jl:66]: Done

File: src/GaussianProcessEmulator.jl (hunk 13 above)
Comment (odunbar) [reply src/GaussianProcessEmulator.jl:324]: Done

**Review by ilopezgp — APPROVED** [review]

**Pull request merged**